### ISSUE TITLE
Return more than just eax as a result in JIT mode

### ISSUE THREAD

**Issue by maurisvh** [jonathanvdc/dfasm#13]
Being able to examine the values of more registers than just `eax` after executing some JIT'ed code would be fun. Is this possible?

**Comment by jonathanvdc** [comment]
Yes and no. We can return as many values as we want by putting them in a stack-allocated struct, and configure the Invoke call on the JIT function to indicate a different return type. We cannot, however, easily "peek" at the values in other registers. These values have long been overwritten by the time control is returned to the assembler.

**Comment by maurisvh** [comment]
Could we write the registers somewhere immediately after code execution? Right now we ask the OS for one chunk of writable/executable memory to run our code in -- suppose this one begins at label B. Can we ask the OS for
- another piece of writable/executable memory -- say at label B --, and
- another piece of readable/writable memory, at label R,

so that we can do something like this:

```
A: call B
   ; registers are intact at this point, so
   write registers to R
   ret

B: ; our code

R: ; an array big enough to hold the registers we wish to examine
```

call A from the assembler, and examine whatever is in R after control returns to it?

**Comment by jonathanvdc** [comment]
That would work. We could even allocate a single block of contiguous memory and use that instead of the thunk/function/array approach. I'd like to go beyond just looking at a few registers, though, as the above technique does have some drawbacks:
- We can only glance at a very small amount of (mostly insignificant) values. The thunk constrains our ability to return a meaningful value. Suppose we wanted to return a struct that contains more int-sized fields than we have registers. Said return struct would have to be stack-allocated, but we lost our ability to properly stack-allocate anything when we used a call instruction, which modifies the stack. A call instruction is also not the only thing you need to call a function in most x64 calling conventions. The Windows calling convention requires "shadow space" on the stack, and AMD deems a "red zone" necessary. Maintaining a thunk for every calling convention and having it manage a separate return struct region on the stack is complicated, and requiring the code we're JITing to use a slightly modified version of whatever calling convention we're using right now is a perversion at best.
- We can't pass any arguments, for the exact same reason. Some of them might be stack-allocated, after all.

More importantly, a JIT is _not_ an emulator. Emulators are powerful debugging tools that allow us to see the contents of every register while stepping through instructions. The point of a JIT is to execute a function, same as an executable file, but without the overhead of object files, linkers and executables. Glancing at "implementation details" such as register values, is not part of a JIT's mission statement.

Right now, the functions we're JITing are all int(void) functions. I'd very much like to see that change. Fortunately, libjit is in fact capable of JITing a function with just about any signature right now.

Adding some extra input parameters for the return and argument types to dfasm would allow us to retrieve _useful_ values from the JITted function. That way, we shift the burden of dealing with calling conventions and processor architectures to the assembly programmer/compiler, and allow us to JIT "real" functions - even a comparatively simple one like Fibonacci has an int(int) signature.

In short, I think a more comprehensive approach to input and output will yield us a far more powerful execution environment than some register-saving scheme.

**Comment by maurisvh** [comment]
Yeah, a generic approach like that would be even cooler! I'm not sure what the limitations are, or how hard it is to pass arguments to the JIT. If you can make libjit handle this in time, maybe it'd be worth considering; it would certainly turn our JIT support from a proof of concept into something more.

**Comment by jonathanvdc** [comment]
libjit can already handle this scenario. `JitFunction.Invoke<T>` is varargs, and `T` specifies the return type. `func.Invoke[int](5)` should work just fine right now. dfasm.py just needs to pass the correct arguments to the JITted function.

**Comment by maurisvh** [comment]
That's neat. Maybe we should test this and close the issue if it works?

**Comment by jonathanvdc** [comment]
I created a hacky single-argument `-arg:` switch. Tested with ipy64 (32-bit ipy uses stdcall, so that wouldn't work for this x86 code).

```
>ipy64 dfasm.py -jit -arg:42
Ready.
mov eax, ecx
ret

42
```

**Comment by jonathanvdc** [comment]
Fibonacci seems to be working. [This commit](https://github.com/jonathanvdc/dfasm/commit/4f1523a2d5d5b4d759ca3421de95d162cfb90460) includes a few JIT examples.
